core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -869,7 +876,20 @@ private[spark] class BlockManager(
           blockInfoManager.unlock(blockId)
         }
       } else {
-        blockInfoManager.removeBlock(blockId)
+        if (exceptionWasThrown) {
```

Comment:
I didn't combine it so that the `logWarning` wouldn't need to be duplicated, but that's not a great rationale.

`removeBlockInternal` (which is used in both this `if` and `else` case now) will handle releasing the lock (this happens in the `blockInfoManager.removeBlock` call).